coverage.py

```diff
@@ -350,83 +351,208 @@ def items_that_need_coverage(self, identifiers, **kwargs):
-        """
-        _db = Session.object_session(output_source)
-        super(CoverageProvider, self).__init__(
-            _db, service_name, operation, batch_size, cutoff_time
-        )
-        if input_identifier_types and not isinstance(input_identifier_types, list):
-            input_identifier_types = [input_identifier_types]
-        self.input_identifier_types = input_identifier_types
-        self.output_source_name = output_source.name
+        super(IdentifierCoverageProvider, self).__init__(_db, **kwargs)
+
+        self.collection = collection
         self.input_identifiers = input_identifiers
+        self.replacement_policy = (
+            replacement_policy or self._default_replacement_policy
+        )
+        
+        if not self.DATA_SOURCE_NAME:
+            raise ValueError(
+                "%s must define DATA_SOURCE_NAME" % self.__class__.__name__
+            )
+        
+        # Get this information immediately so that an error happens immediately
+        # if INPUT_IDENTIFIER_TYPES is not set properly.
+        self.input_identifier_types = self._input_identifier_types()
+
+    @property
+    def _default_replacement_policy(self):
+        """Unless told otherwise, assume that we are getting
+        this data from a reliable metadata source.
+        """
+        return ReplacementPolicy.from_metadata_source()
         
+    @classmethod
+    def _input_identifier_types(cls):
+        """Create a normalized value for `input_identifier_types`
+        based on the INPUT_IDENTIFIER_TYPES class variable.
+        """
+        value = cls.INPUT_IDENTIFIER_TYPES
+
+        # Nip in the bud a situation where someone subclassed this
+        # class without thinking about a value for
+        # INPUT_IDENTIFIER_TYPES.
+        if value is cls.NO_SPECIFIED_TYPES:
+            raise ValueError(
+                "%s must define INPUT_IDENTIFIER_TYPES, even if the value is None." % (cls.__name__)
+            )
+
+        if not value:
+            # We will be processing every single type of identifier in
+            # the system. This (hopefully) means that the identifiers
+            # are restricted in some other way, such as being licensed
+            # to a specific Collection.
+            return None
+        elif not isinstance(value, list):
+            # We will be processing every identifier of a given type.
+            return [value]
+        else:
+            # We will be processing every identify whose type belongs to
+            # a list of types.
+            return value
+
+    @property
+    def data_source(self):
+        """Look up the DataSource object corresponding to the
+        service we're running this data through.
+
+        Out of an excess of caution, we look up the DataSource every
+        time, rather than storing it, in case a CoverageProvider is
+        ever used in an environment where the database session is
+        scoped (e.g. the circulation manager).
```

Comment:
I'm not sure if this is helpful anymore, since we're also storing the `Collection` object with the `CoverageProvider`.